### ISSUE TITLE
Enable overriding the message of a json endpoint through the query string in the URL

### ISSUE THREAD

**Issue by HannesWell** [badges/shields#6815]
:clipboard: **Description**

It is possible to override for example the label of a JSON endpoint in the query string like this:
`https://img.shields.io/endpoint?label=another%20label&url=<URL-to-my-endpoint>.json`

But it does not seem to be possible to override the message of a JSON endpoint through the query string like this:
`https://img.shields.io/endpoint?message=another%20message&url=<URL-to-my-endpoint>.json`
It would be great if overriding the message trough the query string would be possible, too.

We consider to provide a unified batch template by using a JSON endpoint, that can be used by multiple Eclipse projects and where each project only overrides the message with its name in the URL. This way all projects use similar batches with an individual message.

**Comment by paulmelnikow** [comment]
Possibly related to this use case: #6786

**Comment by HannesWell** [comment]
Thanks for the reference, but I have to admit that I'm new to badges and not familiar with the deep details and therefore don't fully understand the conversation.

Our intention why we consider to use an own JSON endpoint is to be able to configure the entire badge, except for the message, at one location. Each Eclipse-project using that badge should only have an individual message, that is expected to be the project name.
In case the badge should change one day (e.g. because Eclipse changes its branding colours) it would only be necessary to modify the JSON file at one location and all badges that use it should change accordingly.

Do you think another approach would be more suitable in our situation?

If you are interested in some more details, you can find them here: https://bugs.eclipse.org/bugs/show_bug.cgi?id=574646

**Comment by paulmelnikow** [comment]
It sounds like the ask is to be able to specify a collection of static badges whose formatting is specified in a JSON file and the message in a query parameter.

I could imagine this being solved with a new endpoint for a collection of static badges, along the lines of the proposal I described in https://github.com/badges/shields/issues/6786#issuecomment-882725827. That would require adding an entry for each project into the JSON file.

However for your use case, I think a better solution would be to check in a script which fetches a list of the projects, generates the static SVGs, and dumps them onto a CDN somewhere – probably in the same place you were thinking of putting that endpoint JSON file.

**Comment by calebcartwright** [comment]
> It sounds like the ask is to be able to specify a collection of static badges whose formatting is specified in a JSON file and the message in a query parameter.

That was my sense as well, with the underlying motivating use case being the desire to be able to centrally manage/bulk update a set of similar-but-not-identical **static** badges used in different repositories.

@HannesWell  - I can understand the desire to automate/simplify the process of updating a bunch of images, but I personally think that capability is outside the remit of Shields; the Endpoint badge is definitely not geared towards that type of use case.

I was going to suggest something very similar to what Paul suggested, with the option of just using GitHub to host the svg files (so the aforementioned script could just check them into source) and then using `<img>` with the respective githubusercontent urls to add the badges to the respective projects.

**Comment by paulmelnikow** [comment]
> with the underlying motivating use case being the desire to be able to centrally manage/bulk update a set of similar-but-not-identical **static** badges used in different repositories.

:+1:

**Comment by HannesWell** [comment]
> It sounds like the ask is to be able to specify a collection of static badges whose formatting is specified in a JSON file and the message in a query parameter.

Yes, we would like to generate badges that all have the same formatting, colour and label and only have an individual message-text.

Thanks for your suggestions.
In principle there is a list of relevant projects that could use the badges and it probably would be possible to generate the entries for a collection of static badges from that list or to generate and store the static svg-files directly.
However being able to modify the message from the query string would just make thinks a lot more easy, but of course the scope of shields is not all-embracing.

I just wonder why it is possible to override the label from the query string, but it is not possible to override the message? From my naive perspective the difference is only that the label is the text on the left and the message the text on the right of a badge.
Naively spoken, if we have decided to put the individual text on the left, it were already possible to accomplish our goal. Or is this only by chance.

**Comment by paulmelnikow** [comment]
The use case for customizing the label is meant for situations where you want to display two CI badges, and need to differentiate them (i.e. they can't both be labeled "build"). In that case there's no need to override the right text.

I think the reason you're finding this mismatch in feature capabilities surprising is that you're using Shields in a way more like what was requested in #2824, for making a custom action button, whereas Shields is primarily designed for rendering status badges. We of course don't mind that you use Shields for these buttons, but at the same time it isn't the primary use case for which Shields was intended.

Thinking more about #2824 and how these buttons have been used in the wild, another design suggestion would be to exclude the package name from the button entirely:

https://img.shields.io/static/v1?logo=eclipseide&label&color=orange&message=Create+Development+Environment&style=for-the-badge&logoColor=white
![](https://img.shields.io/static/v1?logo=eclipseide&label&color=orange&message=Create+Development+Environment&style=for-the-badge&logoColor=white)

Then they're all the same and you can generate it once and pop it on a CDN somewhere.

**Comment by HannesWell** [comment]
Thanks for the explanation. I expected a reason pointing in that direction.

Your design suggestion would be another option. However it would be nicer to 'personalize' the buttons for each project.
But I will see how we will handle this problem in the end.
Anyway, thanks again.

**Comment by JP-Ellis** [comment]
I stumbled on this issue as I was running into the same issue myself.

I was wanting to create a badge that is similar to a lot of status badges where the left side shows the category, and the right side shows information about the category. For example:

![Pypi](https://img.shields.io/pypi/pyversions/pact-python.svg)

In my specific case, I want a family of badges such as:

- ![Pact Badge](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/jp-ellis/pact-foundation-dotgithub/feat/pact-badge/assets/badge/v1.json)
- ![Pact Badge](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/jp-ellis/pact-foundation-dotgithub/feat/pact-badge/assets/badge/v2.json)
- ![Pact Badge](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/jp-ellis/pact-foundation-dotgithub/feat/pact-badge/assets/badge/v3.json)
- ![Pact Badge](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/jp-ellis/pact-foundation-dotgithub/feat/pact-badge/assets/badge/v4.json)

<details>
<summary><i>Screenshot of above if links change in the future</i></summary>
<img width="374" alt="image" src="https://github.com/badges/shields/assets/3196162/22c19b24-ac5b-4d32-927c-531d817a42e4">
</details>

I clearly don't want to create every iteration of the right hand side.

Is this something that endpoints aren't intended to support?

**Comment by bcaddy** [comment]
I ran across this and would also like to be able to change the message in a badge using a query string. The use case for us is that we support many open source projects and want a way to indicate that my team was worked on the project in a given time frame. So the ideal badge would look like: 

[![Princeton RSE Badge](https://img.shields.io/badge/Princeton_RSE-2025-%23F58025.png?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA0AAAAQCAYAAADNo/U5AAAACXBIWXMAAAB4AAAAeAHq8mr2AAAAGXRFWHRTb2Z0d2FyZQB3d3cuaW5rc2NhcGUub3Jnm+48GgAAAfBJREFUKJFtkk9IFGEYxn/fOB87jCKFboUuBa5zykshRH8YWCykQ1RHg71FdA+6SGAdhaCbdIiKLiJBpAlz8hIFEuQh2FgyzRqFWmaV2XZ1dnXfDu7AOvnAd/je5/l97/fCi4iQPEAvMAP0HOofAgwBXwEBCsDpZEYBT4A/gAWcz2QyuXQ63dFoNNBaUyqVdn3fXwAWgQhIA+y2XhXXdcXzPCkUCjI9PS3FYlE8zxPXdSXOAHWjRQMQhiH1ep1KpYJSiiAIKJfLhGFIm3ZMYB1wALTWBEHA5uYm5XKZKIqIogitdTvkq5Gs/bQ7ZQyTUGf3kY5quLWXrIdR85M5l898EeROu1Es1bn95jdT148xdDyVwOSZoczGXGtAAJY2Iq489/m8scPoC5/FXzsHCMPcmzes8ZU1YB7g/Y9trr70CWr7v9rabnLt1ToL32sx9M4aX1lDRKg9cs7NjPWJZSoBJGUquXfpqKRad8tUMjPWJ7WJ7IUDG5EbsFcBsbWS2Xy/VCccmc33i633wdyAvRpnjbjvmUHLdXp04/WtPkayNgAjWZu3+X6cHr07fKLrcpxVrX0D4Of9wYu9nYYHdLUN/7cUNkdPPV7+GBeMNpOTk8sfBHUTqMaAKHWjHfivU6zaQ+csqCkMuWs/+LaU9P8BpwYUBc5MbdwAAAAASUVORK5CYII=&labelColor=%235A575B)](https://researchcomputing.princeton.edu/services/research-software-engineering)

We'd like the year (i.e. message) to be overridable so that each project can have its own info but the rest of the badge to be defined as a JSON endpoint so that if we need to change style/branding/etc later we can do it in a single place and have those changes reflected everywhere.
